File: rust/crates/syntecnia-stdlib/src/cron.rs
//! Cron nativo. Port de `syntecnia/stdlib/cron.py`.
//!
//! Cada job corre en su propio hilo (como `threading.Timer` del oráculo): duerme
//! el intervalo y ejecuta; si es repetitivo, reprograma. La cancelación setea un
//! flag y despierta el hilo (`unpark`) para que salga limpio.
//!
//! Capa 6: la registración (cron_every/list/status/cancel) es lo que se testea. La
//! ejecución de tasks Syntecnia desde un hilo de cron es un tema de concurrencia
//! (diferido, como agents); el builtin del motor registra un task no-op por ahora.

use std::rc::Rc;
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{Duration, Instant};

use indexmap::IndexMap;

use syntecnia_core::interpreter::{Control, Interpreter};
use syntecnia_core::number::py_float_str;
use syntecnia_core::types::{syn_bool, syn_float, syn_int, syn_list, syn_map, syn_text, SynValue};

/// Tarea de un job (thread-safe, corre en el hilo del timer).
pub type Task = Box<dyn Fn() + Send + 'static>;

/// Vista de un job para list_jobs/format_status.
#[derive(Clone, Debug)]
pub struct JobInfo {
    pub name: String,
    pub interval: f64,
    pub repeating: bool,
    pub active: bool,
    pub run_count: u64,
    pub errors: u64,
}

struct JobHandle {
    interval: f64,
    repeating: bool,
    cancelled: Arc<AtomicBool>,
    run_count: Arc<AtomicU64>,
    thread: thread::Thread,
}

/// Scheduler de tareas en background. No bloquea.
pub struct CronScheduler {
    jobs: Mutex<IndexMap<String, JobHandle>>,
}

impl Default for CronScheduler {
    fn default() -> Self {
        Self::new()
    }
}

impl CronScheduler {
    pub fn new() -> Self {
        Self {
            jobs: Mutex::new(IndexMap::new()),
        }
    }

    pub fn every(&self, interval_seconds: f64, name: &str, task: Task) {
        self.schedule(interval_seconds, name, task, true);
    }

    pub fn after(&self, delay_seconds: f64, name: &str, task: Task) {
        self.schedule(delay_seconds, name, task, false);
    }

    fn schedule(&self, interval: f64, name: &str, task: Task, repeating: bool) {
        // Cancela un job existente con el mismo nombre.
        self.cancel(name);

        let cancelled = Arc::new(AtomicBool::new(false));
        let run_count = Arc::new(AtomicU64::new(0));
        let dur = Duration::from_secs_f64(interval.max(0.0));
        let c = cancelled.clone();
        let rc = run_count.clone();

        let jh = thread::spawn(move || loop {
            // Espera `dur`, despertable por unpark (cancelación), robusto a wakeups espurios.
            let deadline = Instant::now() + dur;
            loop {
                if c.load(Ordering::SeqCst) {
                    return;
                }
                let now = Instant::now();
                if now >= deadline {
                    break;
                }
                thread::park_timeout(deadline - now);
            }
            if c.load(Ordering::SeqCst) {
                return;
            }
            rc.fetch_add(1, Ordering::SeqCst);
            task();
            if !repeating {
                return;
            }
        });

        let thread = jh.thread().clone(); // jh se dropea → hilo desacoplado (detached)
        self.jobs.lock().unwrap().insert(
            name.to_string(),
            JobHandle {
                interval,
                repeating,
                cancelled,
                run_count,
                thread,
            },
        );
    }

    pub fn cancel(&self, name: &str) -> bool {
        if let Some(job) = self.jobs.lock().unwrap().shift_remove(name) {
            job.cancelled.store(true, Ordering::SeqCst);
            job.thread.unpark();
            true
        } else {
            false
        }
    }

    pub fn cancel_all(&self) {
        let mut jobs = self.jobs.lock().unwrap();
        for (_, job) in jobs.drain(..) {
            job.cancelled.store(true, Ordering::SeqCst);
            job.thread.unpark();
        }
    }

    pub fn list_jobs(&self) -> Vec<JobInfo> {
        self.jobs
            .lock()
            .unwrap()
            .iter()
            .map(|(name, j)| JobInfo {
                name: name.clone(),
                interval: j.interval,
                repeating: j.repeating,
                active: true,
                run_count: j.run_count.load(Ordering::SeqCst),
                errors: 0,
            })
            .collect()
    }

    pub fn format_status(&self) -> String {
        let jobs = self.list_jobs();
        if jobs.is_empty() {
            return "No scheduled tasks.".to_string();
        }
        let mut lines = vec![format!("Scheduled Tasks ({}):", jobs.len())];
        for j in &jobs {
            let repeat = if j.repeating {
                format!("every {}s", py_float_str(j.interval))
            } else {
                "once".to_string()
            };
            let status = if j.active { "active" } else { "cancelled" };
            lines.push(format!(
                "  [{}] {}: {}, runs: {}, errors: {}",
                status, j.name, repeat, j.run_count, j.errors
            ));
        }
        lines.join("\n")
    }
}

impl Drop for CronScheduler {
    fn drop(&mut self) {
        self.cancel_all();
    }
}

// -- Builtins --

fn raw_str(v: &SynValue) -> String {
    match v {
        SynValue::Text(s) => s.to_string(),
        SynValue::Number(n) => n.to_string(),
        SynValue::Bool(b) => if *b { "True" } else { "False" }.to_string(),
        SynValue::Nothing => "None".to_string(),
        other => other.to_string(),
    }
}

fn arg_f64(v: &SynValue) -> f64 {
    match v {
        SynValue::Number(n) => n.to_f64(),
        SynValue::Text(s) => s.trim().parse::<f64>().unwrap_or(0.0),
        _ => 0.0,
    }
}

/// Nombre del task (de un SynTaskValue/Builtin) o str del valor.
fn task_name(v: &SynValue) -> String {
    match v {
        SynValue::Task(t) => t.name.clone(),
        SynValue::Builtin(b) => b.name.clone(),
        other => raw_str(other),
    }
}

fn err(msg: &str) -> Control {
    Control::Error(syntecnia_core::interpreter::RuntimeError::new(msg))
}

/// Registra los builtins de cron, compartiendo el scheduler.
pub fn register_cron_builtins(interp: &Interpreter, scheduler: Rc<CronScheduler>) {
    // cron_every(seconds, task) — registra (la ejecución del task es diferida).
    {
        let sched = scheduler.clone();
        interp.register_builtin(
            "cron_every",
            2,
            Rc::new(move |_i, args, _loc| {
                let interval = arg_f64(args.first().ok_or_else(|| err("missing argument"))?);
                let name = task_name(args.get(1).ok_or_else(|| err("missing argument"))?);
                sched.every(interval, &name, Box::new(|| {}));
                Ok(syn_text(name))
            }),
        );
    }

    // cron_after(seconds, task)
    {
        let sched = scheduler.clone();
        interp.register_builtin(
            "cron_after",
            2,
            Rc::new(move |_i, args, _loc| {
                let delay = arg_f64(args.first().ok_or_else(|| err("missing argument"))?);
                let name = task_name(args.get(1).ok_or_else(|| err("missing argument"))?);
                sched.after(delay, &name, Box::new(|| {}));
                Ok(syn_text(name))
            }),
        );
    }

    // cron_cancel(name) → bool
    {
        let sched = scheduler.clone();
        interp.register_builtin(
            "cron_cancel",
            1,
            Rc::new(move |_i, args, _loc| {
                let name = raw_str(args.first().ok_or_else(|| err("missing argument"))?);
                Ok(syn_bool(sched.cancel(&name)))
            }),
        );
    }

    // cron_list() → lista de maps {name, interval, repeating, active, run_count}
    {
        let sched = scheduler.clone();
        interp.register_builtin(
            "cron_list",
            0,
            Rc::new(move |_i, _args, _loc| {
                let result: Vec<SynValue> = sched
                    .list_jobs()
                    .into_iter()
                    .map(|j| {
                        let mut m = IndexMap::new();
                        m.insert("name".to_string(), syn_text(j.name.as_str()));
                        m.insert("interval".to_string(), syn_float(j.interval));
                        m.insert("repeating".to_string(), syn_bool(j.repeating));
                        m.insert("active".to_string(), syn_bool(j.active));
                        m.insert("run_count".to_string(), syn_int(j.run_count as i64));
                        syn_map(m)
                    })
                    .collect();
                Ok(syn_list(result))
            }),
        );
    }

    // cron_status() → texto
    {
        let sched = scheduler.clone();
        interp.register_builtin(
            "cron_status",
            0,
            Rc::new(move |_i, _args, _loc| Ok(syn_text(sched.format_status()))),
        );
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::atomic::AtomicUsize;

    #[test]
    fn cron_scheduler_basic() {
        let sched = CronScheduler::new();
        let counter = Arc::new(AtomicUsize::new(0));
        let c = counter.clone();
        sched.every(0.1, "counter", Box::new(move || {
            c.fetch_add(1, Ordering::SeqCst);
        }));
        thread::sleep(Duration::from_millis(350));
        sched.cancel("counter");
        assert!(counter.load(Ordering::SeqCst) >= 2);
    }

    #[test]
    fn cron_after_runs_once() {
        let sched = CronScheduler::new();
        let result = Arc::new(Mutex::new(None::<String>));
        let r = result.clone();
        sched.after(0.1, "delayed", Box::new(move || {
            *r.lock().unwrap() = Some("done".to_string());
        }));
        thread::sleep(Duration::from_millis(300));
        assert_eq!(*result.lock().unwrap(), Some("done".to_string()));
        *result.lock().unwrap() = None;
        thread::sleep(Duration::from_millis(200));
        assert_eq!(*result.lock().unwrap(), None); // no repite
    }

    #[test]
    fn cron_cancel_stops() {
        let sched = CronScheduler::new();
        let counter = Arc::new(AtomicUsize::new(0));
        let c = counter.clone();
        sched.every(0.1, "test", Box::new(move || {
            c.fetch_add(1, Ordering::SeqCst);
        }));
        thread::sleep(Duration::from_millis(150));
        sched.cancel("test");
        let at_cancel = counter.load(Ordering::SeqCst);
        thread::sleep(Duration::from_millis(300));
        assert_eq!(counter.load(Ordering::SeqCst), at_cancel);
    }

    #[test]
    fn cron_list_jobs() {
        let sched = CronScheduler::new();
        sched.every(60.0, "job1", Box::new(|| {}));
        sched.every(120.0, "job2", Box::new(|| {}));
        let jobs = sched.list_jobs();
        assert_eq!(jobs.len(), 2);
        let names: Vec<&str> = jobs.iter().map(|j| j.name.as_str()).collect();
        assert!(names.contains(&"job1"));
        assert!(names.contains(&"job2"));
        sched.cancel_all();
    }
}

File: synsema/stdlib/cron.py
"""
Synsema Native Cron — Scheduled task execution.

Schedule tasks to run at intervals or specific times:

    cron_every(30, sync_inventory)              -- every 30 seconds
    cron_every(3600, generate_report)           -- every hour
    cron_after(10, send_notification)           -- once, after 10 seconds
    cron_cancel("sync_inventory")               -- cancel a scheduled task
    cron_list()                                 -- show all scheduled tasks

Each scheduled task runs in its own thread. The cron scheduler
runs as a background daemon thread — it doesn't block the program.

Uses only Python stdlib (threading.Timer). No APScheduler, no celery.
"""

import threading
import time
from typing import Dict, List, Optional, Callable, Any
from dataclasses import dataclass, field


@dataclass
class CronJob:
    """A scheduled job."""
    name: str
    interval_seconds: float
    task_fn: Callable
    repeating: bool = True
    active: bool = True
    run_count: int = 0
    last_run: float = 0
    created_at: float = 0
    timer: Optional[threading.Timer] = field(default=None, repr=False)
    errors: List[str] = field(default_factory=list)

    def __post_init__(self):
        if not self.created_at:
            self.created_at = time.time()


class CronScheduler:
    """
    Background task scheduler for Synsema.

    Manages recurring and one-shot scheduled tasks.
    Each task runs in its own thread. The scheduler is non-blocking.
    """

    def __init__(self):
        self.jobs: Dict[str, CronJob] = {}
        self._lock = threading.Lock()

    def every(self, interval_seconds: float, name: str,
              task_fn: Callable) -> CronJob:
        """Schedule a repeating task."""
        with self._lock:
            # Cancel existing job with same name
            if name in self.jobs:
                self._cancel_job(self.jobs[name])

            job = CronJob(
                name=name,
                interval_seconds=interval_seconds,
                task_fn=task_fn,
                repeating=True,
            )
            self.jobs[name] = job
            self._schedule_next(job)
            return job

    def after(self, delay_seconds: float, name: str,
              task_fn: Callable) -> CronJob:
        """Schedule a one-shot task after a delay."""
        with self._lock:
            if name in self.jobs:
                self._cancel_job(self.jobs[name])

            job = CronJob(
                name=name,
                interval_seconds=delay_seconds,
                task_fn=task_fn,
                repeating=False,
            )
            self.jobs[name] = job
            self._schedule_next(job)
            return job

    def cancel(self, name: str) -> bool:
        """Cancel a scheduled job."""
        with self._lock:
            if name in self.jobs:
                self._cancel_job(self.jobs[name])
                del self.jobs[name]
                return True
            return False

    def cancel_all(self):
        """Cancel all jobs."""
        with self._lock:
            for job in self.jobs.values():
                self._cancel_job(job)
            self.jobs.clear()

    def list_jobs(self) -> List[Dict[str, Any]]:
        """List all active jobs."""
        with self._lock:
            result = []
            for job in self.jobs.values():
                result.append({
                    "name": job.name,
                    "interval": job.interval_seconds,
                    "repeating": job.repeating,
                    "active": job.active,
                    "run_count": job.run_count,
                    "last_run": job.last_run,
                    "errors": len(job.errors),
                })
            return result

    def _schedule_next(self, job: CronJob):
        """Schedule the next execution of a job."""
        if not job.active:
            return

        def run():
            job.last_run = time.time()
            job.run_count += 1
            try:
                job.task_fn()
            except Exception as e:
                job.errors.append(f"Run {job.run_count}: {e}")
            # Reschedule if repeating
            if job.repeating and job.active:
                self._schedule_next(job)

        timer = threading.Timer(job.interval_seconds, run)
        timer.daemon = True
        job.timer = timer
        timer.start()

    def _cancel_job(self, job: CronJob):
        job.active = False
        if job.timer:
            job.timer.cancel()

    def format_status(self) -> str:
        jobs = self.list_jobs()
        if not jobs:
            return "No scheduled tasks."
        lines = [f"Scheduled Tasks ({len(jobs)}):"]
        for j in jobs:
            repeat = f"every {j['interval']}s" if j['repeating'] else "once"
            status = "active" if j['active'] else "cancelled"
            lines.append(
                f"  [{status}] {j['name']}: {repeat}, "
                f"runs: {j['run_count']}, errors: {j['errors']}"
            )
        return "\n".join(lines)


def register_cron_builtins(env, scheduler: CronScheduler, interpreter=None,
                           live_output: bool = True):
    """Register cron builtins in a Synsema environment."""
    import sys as _sys
    from ..core.types import (
        SynValue, BuiltinTask, SynTask, SynTaskValue,
        syn_number, syn_text, syn_bool, syn_nothing, syn_list, syn_map,
    )

    # For cron tasks, output goes directly to stdout (not buffered)
    _original_output = interpreter.output_callback if interpreter else None

    def _cron_output(text):
        """Route cron task output to stdout in real time."""
        if _original_output:
            _original_output(text)
        # Also print to real stdout for serve/daemon mode
        print(text, flush=True)

    def _cron_every(args):
        """cron_every(seconds, task_or_name)"""
        interval = float(args[0].raw)
        task_val = args[1]

        # Get task name and callable
        if isinstance(task_val.raw, SynTaskValue):
            name = task_val.raw.name
            def run():
                if interpreter:
                    saved_cb = interpreter.output_callback
                    interpreter.output_callback = _cron_output
                    try:
                        interpreter._call_value(task_val, [], None)
                    finally:
                        interpreter.output_callback = saved_cb
        elif isinstance(task_val.raw, BuiltinTask):
            name = task_val.raw.name
            def run():
                task_val.raw.func([])
        else:
            name = str(task_val.raw)
            def run():
                pass

        scheduler.every(interval, name, run)
        return syn_text(name)

    def _cron_after(args):
        """cron_after(seconds, task_or_name)"""
        delay = float(args[0].raw)
        task_val = args[1]

        if isinstance(task_val.raw, SynTaskValue):
            name = task_val.raw.name
            def run():
                if interpreter:
                    interpreter._call_value(task_val, [], None)
        elif isinstance(task_val.raw, BuiltinTask):
            name = task_val.raw.name
            def run():
                task_val.raw.func([])
        else:
            name = str(task_val.raw)
            def run():
                pass

        scheduler.after(delay, name, run)
        return syn_text(name)

    def _cron_cancel(args):
        """cron_cancel(name)"""
        name = str(args[0].raw)
        return syn_bool(scheduler.cancel(name))

    def _cron_list(args):
        """cron_list() → list of job info maps"""
        jobs = scheduler.list_jobs()
        result = []
        for j in jobs:
            result.append(syn_map({
                "name": syn_text(j["name"]),
                "interval": syn_number(j["interval"]),
                "repeating": syn_bool(j["repeating"]),
                "active": syn_bool(j["active"]),
                "run_count": syn_number(j["run_count"]),
            }))
        return syn_list(result)

    def _cron_status(args):
        """cron_status() → formatted text"""
        return syn_text(scheduler.format_status())

    builtins = {
        "cron_every": BuiltinTask("cron_every", _cron_every, 2),
        "cron_after": BuiltinTask("cron_after", _cron_after, 2),
        "cron_cancel": BuiltinTask("cron_cancel", _cron_cancel, 1),
        "cron_list": BuiltinTask("cron_list", _cron_list, 0),
        "cron_status": BuiltinTask("cron_status", _cron_status, 0),
    }
    for name, builtin in builtins.items():
        env.set(name, SynValue(raw=builtin, type=SynTask()))
